File: 0066-plus-one/0066-plus-one.py
class Solution:
    """
    def plusOne(self, digits: List[int]) -> List[int]:
        for i in range(len(digits)-1, -1, -1):
            if digits[i] == 9:
                digits[i] = 0
            else:
                digits[i] += 1
                return digits
        
        if digits[0] == 0:
            digits.insert(0, 1)
            
        return digits   
    """    
    def plusOne(self, digits: List[int]) -> List[int]:
        if digits[-1] == 9:
            if len(digits) == 1:
                return [1, 0]
            
            return self.plusOne(digits[:-1]) + [0]
        
        digits[-1] += 1
        return digits
        
             

File: 0066-plus-one/0066-plus-one.c
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* plusOne(int* digits, int digitsSize, int* returnSize){
    int digit;
    int carry = 1;
    int* sum = malloc(sizeof(int) * (digitsSize+1));
    
    for (int i = digitsSize-1; i >= 0; i--) {
        digit = digits[i] + carry;
        if (digit > 9) {
            carry = 1;
            digit -= 10;
        } else {
            carry = 0;
        }
        sum[i+1] = digit;
    }
    
    if (carry) {
        *returnSize = digitsSize + 1;
        sum[0] = carry;
    } else {
        *returnSize = digitsSize;
        // shift all elements left
        for (int i = 1; i < digitsSize+1; i++) {
            sum[i-1] = sum[i];
        }
    }
    
    return sum;
}